### PR TITLE
cpu: aarch64: add JIT implementation for eltwise_f32

### PR DESCRIPTION
**Pull request by kawakami-k**
# Description

This PR adds JIT support of eltwise f32 for AArch64.
It includes JIT implementation for SVE 512.
 [Related RFC#841](https://github.com/oneapi-src/oneDNN/pull/841) 
 [Related PR#850](https://github.com/oneapi-src/oneDNN/pull/850)

At this time, pow and sigmoid, recently implemented for x64, are not supported. `status_t jit_uni_eltwise_fwd_t<isa, d_type>::pd_t::init(engine_t *engine)` returns `status::unimplemented;`, if either pow or sigmoid is specified. 

# Checklist

## Code-change submissions

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally?
- [x] Have you formatted the code using clang-format?

### New features

- [N/A] Have you added relevant tests?
- [N/A] Have you provided motivation for adding a new feature?

There is no need to add a new tests for this PR, 
since we can use the existing gtests and benchdnn for this PR.

### CODE REVIEW

**Comment by vpirogov** [comment]
Hi @kawakami-k,

I'm happy to see you progressing with JIT optimizations. Just wanted to let you know that majority of the core team engineers are enjoying the holiday season and time off, so we might be slow with code review until Jan 4th next year. Code cutoff and release date for [oneDNN v2.1](https://github.com/oneapi-src/oneDNN/milestone/8) were pushed a few weeks out, so there's time for these changes to land on time.

**Comment by kawakami-k** [comment]
Thank you for sharing your situation.
We are also enjoyning the holidays until Jan 4th next year.
Have a good X'mas and New Year's holiday!

File: src/cpu/aarch64/injectors/injector_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/injector_utils.hpp:3]:
```suggestion
* Copyright 2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp [hunk 2]
```diff
@@ -0,0 +1,2253 @@
+/*******************************************************************************
+* Copyright 2019-2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:3]:
```suggestion
* Copyright 2019-2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp [hunk 3]
```diff
@@ -0,0 +1,2253 @@
...
+        const injector_utils::vmm_index_set_t &vmm_idxs) {
+    using namespace alg_kind;
+    using namespace Xbyak_aarch64::util;
+    preserved_vecs_count = 0;
+    vecs_to_preserve = aux_vecs_count();
+    const auto start_idx = *(vmm_idxs.begin());
+    const auto end_idx = *(vmm_idxs.rbegin()) + 1;
+    start_idx_tail = vmm_idxs.begin();
+
+    // For sse41 mask register has to be Xmm(0)
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:45]: does this restriction applies to aarch64?
In any case, please modify the comment to not mention sse and Xmms on aarch64.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp [hunk 4]
```diff
@@ -0,0 +1,2253 @@
...
+    vmm_aux2 = TRegS(preserved_vec_idxs[3]);
+    vmm_aux3 = TRegS(preserved_vec_idxs[4]);
+    vmm_aux4 = TRegS(preserved_vec_idxs[5]);
+    vmm_aux5 = TRegS(preserved_vec_idxs[6]);
+    vmm_aux6 = TRegS(preserved_vec_idxs[7]);
+    vmm_aux7 = TRegS(preserved_vec_idxs[8]);
+}
+
+// Uses injector masks objects: k_mask (>= avx512_common) or vmm_mask (<= avx2).
+// Stores a mask by applying cmpps on two inputs w/ a given predicate.
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:215]: Same as above, please modify comments to reflect the situation on aarch64, not on x64.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp [hunk 5]
```diff
@@ -0,0 +1,2253 @@
...
+        case TRUE_UQ:
+        case UNORD_S:
+        case ORD_S:
+        case FALSE_OS:
+        case TRUE_US:
+        default: assert(!"Unsupported compare mode"); break;
+    }
+}
+
+// Uses injector masks objects: k_mask (>= avx512_common) or vmm_mask (<= avx2).
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:366]: This comment seems irrelevant as you always use `sel` and not masked execution.

File: src/cpu/aarch64/jit_uni_eltwise.cpp [hunk 6]
```diff
@@ -0,0 +1,344 @@
+/*******************************************************************************
+* Copyright 2017-2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/aarch64/jit_uni_eltwise.cpp:3]:
```suggestion
* Copyright 2017-2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

File: src/cpu/cpu_eltwise_list.cpp [hunk 7]
```diff
@@ -1,5 +1,6 @@
 /*******************************************************************************
 * Copyright 2019-2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/cpu_eltwise_list.cpp:3]:
```suggestion
* Copyright 2019-2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp [hunk 8]
```diff
@@ -0,0 +1,2253 @@
...
+    using namespace Xbyak_aarch64::util;
+    const int tanh_n_polynomials = 32;
+
+    // register mapping
+    // TODO: put sign on stack and alias zmm_table2 with vmm_sign to save a reg ?
+    TRegS vmm_dst = vmm_aux1, vmm_src_shift = vmm_aux1, vmm_coeff = vmm_aux1,
+          vmm_pol = vmm_aux2, vmm_indices = vmm_aux3,
+          vmm_src_original = vmm_aux4, vmm_sign = vmm_aux4;
+
+    auto vpermt2ps_aarch64 =
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:499]: I don't know if `vpermt2ps` makes sense in the aarch64 world :-)
Since it just does a table lookup of 64 entries in 2 zreg vectors with f32 entries, you could use `zreg_lut_64_f32` or `lut_zreg2_f32`.

File: src/cpu/aarch64/jit_uni_eltwise.hpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
+/*******************************************************************************
+* Copyright 2017-2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/aarch64/jit_uni_eltwise.hpp:3]:
```suggestion
* Copyright 2017-2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.hpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
+/*******************************************************************************
+* Copyright 2019-2020 Intel Corporation
+* Copyright 2020 FUJITSU LIMITED
```
Comment (mgouicem) [inline src/cpu/aarch64/injectors/jit_uni_eltwise_injector.hpp:3]:
```suggestion
* Copyright 2019-2021 Intel Corporation
* Copyright 2021 FUJITSU LIMITED
```

**Review by mgouicem — APPROVED** [review]
Looks good to me. Thanks for the contribution!

File: src/cpu/aarch64/injectors/injector_utils.hpp (hunk 1 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/injector_utils.hpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp (hunk 2 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/jit_uni_eltwise.cpp (hunk 6 above)
Comment (kawakami-k) [reply src/cpu/aarch64/jit_uni_eltwise.cpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp (hunk 2 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/jit_uni_eltwise.cpp (hunk 6 above)
Comment (kawakami-k) [reply src/cpu/aarch64/jit_uni_eltwise.cpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/cpu_eltwise_list.cpp (hunk 7 above)
Comment (kawakami-k) [reply src/cpu/cpu_eltwise_list.cpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/jit_uni_eltwise.hpp (hunk 9 above)
Comment (kawakami-k) [reply src/cpu/aarch64/jit_uni_eltwise.hpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.hpp (hunk 10 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.hpp:3]: Thank you suggestion. I'll fix it.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp (hunk 3 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:45]: At the moment, aarch64 eltwise supports only sve_512, and these lines are not required.
I'll delete it.

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp (hunk 5 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:366]: Thank you for the comment. I'll modify comments to reflect the situation on aarch64.
(AArch64 SEL corresponds to x64 VBLENDVPS.)

File: src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp (hunk 8 above)
Comment (kawakami-k) [reply src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp:499]: Thank you for the suggestion. I'll use `lut_zreg2_f32`.

**Comment by kawakami-k** [comment]
@mgouicem Thank you for the review. I modified the source code based on the comments and suggestions you gave me.
